Allow to set a desired update interval for camera_proxy_stream view

NovapaX opened the pull request:
## Description:

Set desired image update interval using 'interval' query parameter. defaults to (existing) 0.5 seconds. Harcoded mininum of 0.05 seconds. Selecting a shorter interval will increase disk I/O (comparing image file, but not more network traffic)
Maybe the minimum interval is too short, I don't really know, please discuss!

In the front-end this could be used like:
```
this.cameraFeedSrc = '/api/camera_proxy_stream/' + this.stateObj.entity_id +
             '?interval=10&token=' + this.stateObj.attributes.access_token;
```

note: if the 'interval' parameter is omitted it will use the default (0.5) seconds.

Main use case: convert the frontend camera card to use the camera proxy (with a long interval, or maybe even selectable) and only update the image if it is changed, instead of a new request (and image transfer) every 10 seconds.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

NovapaX committed:
allow to set a desired update interval for camera_proxy_stream view

NovapaX committed:
lint

OttoWinter commented on homeassistant/components/camera/__init__.py:
```diff
@@ -275,6 +275,12 @@ def handle_async_mjpeg_stream(self, request):
         last_image = None
 
+        interval = .5  # default stream framerate: 2 fps
+        min_interval = .05  # max stream framerate: 20 fps
```
I think this should be moved to a global constant.

NovapaX replied:
yes, you are right

OttoWinter commented on homeassistant/components/camera/__init__.py:
```diff
@@ -275,6 +275,12 @@ def handle_async_mjpeg_stream(self, request):
+        min_interval = .05  # max stream framerate: 20 fps
+
+        if request.query.get('interval'):
+            interval = max(float(request.query.get('interval')), min_interval)
```
Let's please not use max here, instead let's raise an exception, render a 500 or something like that to indicate wrong input.

I'm sure some users will be confused when their interval isn't accepted by the server. Some users will definitely try to get 30fps.

Also, I feel like the request parameter could also be called FPS, since that's more the standard of cameras.

NovapaX replied:
Makes sense. And then just use `1/fps` as parameter for `asyncio.sleep()`

NovapaX replied:
Yes, we could raise a `403 Forbidden` with a description of the reason.
In the frontend we would limit what users can select to valid framerate.

OttoWinter replied:
Hmmm, Maybe fps is not too good after all: If fetching the latest image + sending it takes a long time, we would always have *a little bit less* than the fps. That could also lead to confusion.

So do we just leave that limitation on there or track time by ourselves to achieve "true" fps...?

NovapaX replied:
I'd think we can just leave the limitation. It's not a video player.... achieving 'true fps' is not really the point.

NovapaX replied:
I believe INTERVAL would be a better description of the behaviour here.
Users who will figure out how to get to the stream, will be able to understand what an interval is and how that relates to 'fps' I guess.

OttoWinter replied:
Yes, I agree.

NovapaX commented:
Ok. I've been giving this some thought. The feature itself is useful (streaming with a custom interval)
My implementation is sort of taking the wrong approach.
I've got to extend/change the view logic, maybe add a handle_async_image_stream, but not change the handle_async_mjpeg_stream because a lot of camera platforms implement that method themselves where they literally proxy the stream from the camera.... Going to refactor now

NovapaX committed:
refactor into a seperate method.
Keep the handle_async_mjpeg_stream method to be overridden by platforms
so they can keep proxying the direct streams from the camera

NovapaX committed:
change descriptions

NovapaX committed:
consolidate

NovapaX committed:
lint

NovapaX committed:
travis

balloob commented:
I don't think that we should merge this. There are already some camera's that convert streams using FFMPEG which I think is the way to go. For us to create an MJPEG stream on the fly, should be considered a fallback and we should not make it too good, so that people improve the integrations 😉 I also think that it's too easy to eat up all CPU or hammer the camera device if we allow it to be set too high.

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -277,29 +279,39 @@ def handle_async_mjpeg_stream(self, request):
                     last_image = img_bytes
 
-                yield from asyncio.sleep(.5)
+                await asyncio.sleep(interval)
```
This interval does not work as advertised. It's the interval that we wait between finishing writing a frame and before we start fetching the next frame. What people assume when they see the word interval is that it is the time between the start of writing two frames. (note that currently it is not configurable and thus everything worked as advertised)

NovapaX replied:
I explicitly picked interval, and not ‘refresh rate’ or ‘FPS’ because I was already aware of this.

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -413,5 +425,16 @@ class CameraMjpegStream(CameraView):
-        yield from camera.handle_async_mjpeg_stream(request)
+        """Serve camera stream, possibly with interval."""
+        if request.query.get('interval'):
+            """Compose camera stream from stills."""
```
This is not how comments work inside code. Use `# <comment>`

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -413,5 +425,16 @@ class CameraMjpegStream(CameraView):
+            yield from camera.handle_async_still_stream(request, interval)
+        else:
+            """Serve camera stream."""
+            yield from camera.handle_async_mjpeg_stream(request)
```
We've migrated to `async`/`await` syntax. Don't add new `yield from`

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -413,5 +425,16 @@ class CameraMjpegStream(CameraView):
+            """Compose camera stream from stills."""
+            interval = float(request.query.get('interval'))
+            if interval < MIN_STREAM_INTERVAL:
+                return web.Response(status=403,
```
This is not a 403, it's a bad request (400)

NovapaX replied:
You are probably right, I think I just misunderstood the 403 description.

NovapaX commented:
I definitely agree on your point that this is a fallback. And platforms should really provide a proper stream. And if we make it too good it will get too easy to fall back to this resource heavy way of streaming.

However currently we refresh the frontend card with a new request and new download every 10 seconds. The point of this was to convert it so it could (always) use a stream, and maybe pick a faster (or slower interval) it would however only take the necessary bandwidth as new data only comes in when there is new data.

What if we limit it to the current update interval of 0.5 seconds? That’s currently what platforms fall back to if they not implement a proper proxy stream.
I wouldn’t even mind to set it to 1 or 2 seconds. I just care for a more efficient stream to the cams card.

NovapaX committed:
async/await and force min stream interval for fallback stream.

balloob commented:
I was thinking for the preview in the overview to actually add a new method to camera's to be able to provide a thumbnail picture. Right now we stream a lot of data to the frontend.

balloob commented:
Oh I see now how we would be saving data, it's because we would not have to redownload a picture if it has not changed in 10 seconds.

NovapaX commented:
Yes, and we can give a little control in the frontend about the update interval. (no values, maybe just a "tortoise and a hare" switch)

This all started because rpi_camera has a configurable interval. Someone wanted to set it to 30 seconds and 'sync' that interval to the frontend to save some requests and data.
So it's definitely not intended to provide high-fps streaming.

I don't know how much overhead every https request adds vs. an open socket, but with ssl that could add up too I guess?

balloob commented:
I would not bother about the overhead of a request vs a socket. You have convinced me that this is a useful addition 👍

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -411,7 +426,17 @@ class CameraMjpegStream(CameraView):
-        yield from camera.handle_async_mjpeg_stream(request)
+    async def handle(self, request, camera):
+        """Serve camera stream, possibly with interval."""
+        if request.query.get('interval'):
```
let's use a guard clause.

```python
interval = request.query.get('interval')
if interval is None:
    await camera.handle_async_mjpeg_stream(request)
    return

try:
    interval = float(interval)
# etc
```
```

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -411,7 +426,17 @@ class CameraMjpegStream(CameraView):
+            interval = float(request.query.get('interval'))
+            if interval < MIN_STREAM_INTERVAL:
+                return web.Response(status=400,
+                                    reason="Interval must be a number \
```
Reason should be the status code. I think that you're looking for something like `self.json_message`

NovapaX committed:
guard clause. Let the method raise error on interval.

NovapaX committed:
is is not =

NovapaX committed:
what to except when you're excepting

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -411,7 +426,19 @@ class CameraMjpegStream(CameraView):
+        except (ValueError, HomeAssistantError):
+            return web.Response(status=400)
+
+        return web.Response(status=500)
```
This is unreachable, you have a return inside the try and inside the except

NovapaX replied:
Wouldn't we reach it when an error is raised other than the one we catch?

balloob replied:
No, uncaught exceptions blow up the whole function (which is fine, it will trigger a 500)

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -411,7 +426,19 @@ class CameraMjpegStream(CameraView):
+            interval = float(request.query.get('interval'))
+            await camera.handle_async_still_stream(request, interval)
+            return
+        except (ValueError, HomeAssistantError):
```
When do we throw HomeAssistantError ?

NovapaX replied:
It's inside `handle_async_still_stream(request, interval)` on a too low value of interval. But that should probably also just be a ValueError?

balloob replied:
Yeah

NovapaX committed:
raise ValueError, remove unnecessary 500 response

OttoWinter reviewed (commented):
The code looks good to me 👍 Looking forward to this.

Tests might be good though.
